src/nats/connection.cr

```diff
@@ -541,5 +538,9 @@ module NATS
       end
       @socket << "\r\n"
     end
+
+    private def send_flush
+      @flush.send(true) if @flush.@queue.not_nil!.empty?
```

Comment:
Why did they remove simple empty?

Reply:
I asked them about it because I was using that in one of my own shards. Apparently they didn't think channels were supposed to be used for that due to race conditions. For example, if you were breaking from a loop when a channel is drained, but it may only be drained because of the order of scheduling the fibers or due to an I/O blip on the producer side of the channel.

This code doesn't need to worry about the race condition, but this was the workaround I've been using in my own shard and it's worked okay since `Channel::Unbuffered` and `Channel::Buffered` were combined and the method was removed.

Reply:
I was seeing a reduction locally from 700k publishes/sec down to around 50k when I removed the conditional, presumably because `send_flush` kept blocking when it outpaced the `outbound` fiber.

I wonder if there's another way to signal the flush instead of through a `Channel`. It's almost like we need something like a `ConditionVariable` (worst-named concept in computer science, imo 😂) to signal the fiber to wake up, but not necessarily immediately like would happen with `@outbound.resume`. I'll play around with some ideas and see if I can come up with something that is both performant and concurrency-friendly.

Reply:
We could use the trick I used in the Rust client a bit. Gist is if the kicker action is too fast we can wait for a bit as long as things are changing.

So what I did was has a ramp up of times to wait. When kicked I would sleep first one, very short. If when I came out of that the number of messages waiting was the same I would flush to the socket, if not and we were under a minimum of size (64k) and time (10ms) then would wait again with next value that was larger, say 2X or so..

Reply:
Oh interesting, like exponential backoff to automatically adjust to the frequency of messages?

Reply:
Well just if we are sending messages very rapidly do not want the send side to the socket essentially doing a poor man's spin lock so we release lock and sleep for a bit and that can get larger but does not have to..

Reply:
Experimented a bit more this evening. What if the `outbound` fiber flushed every X ms, no matter what? It seems to be working well for me in my local checkout. I was trying to see if publishing unnecessarily would have an effect on performance, but [`IO::Buffered#flush` has a guard around the `write` syscall](https://github.com/crystal-lang/crystal/blob/612825a53c831ce7d17368c8211342b199ca02ff/src/io/buffered.cr#L191-L196) and [`Socket#unbuffered_flush` is a no-op](https://github.com/crystal-lang/crystal/blob/612825a53c831ce7d17368c8211342b199ca02ff/src/socket.cr#L578-L580).

Testing with 1ms and 10ms sleeps in the loop doesn’t have any meaningful effect on publish throughput. It looks like we can just flush the socket whenever we need. The mutex lock in `NATS::Connection#flush_outbound` is probably the most intense operation in that loop if the buffer is empty.

I can go ahead and push that change up here so you can try it out with some of your own tooling if you’d like to test it out. I’m still able to reach 12.8M publishes/sec in a microbenchmark on my machine, which is what I get with this branch in its current state. I’m pleasantly surprised at just how well Crystal has us covered here.

Reply:
Prefer not to spend cpu if we don't have to tbh.. The kick did not work correctly with the backoff?

Reply:
That's what I mean, I found I didn't need to do anything clever to achieve the same throughput. As this PR sits now, I get 12,703,426 publishes/sec averaged out over a minute. By simply flushing every 10ms, I was able to get 12,802,899.

Also, is `nats-server` multithreaded? If not, it seems to be my bottleneck here.

<img width="253" alt="Screenshot of CPU consumption between nats-server and my benchmark client" src="https://user-images.githubusercontent.com/108205/77594978-506f2100-6ece-11ea-8457-aaaaf95a2586.png">

Reply:
nats-server is multi runtime context (via Go) but will be only one processing the inbound socket from the client. 

My issue with 10ms is that if i send a single message we have now introduced 10ms to sending that single message, which is not good. 

My thought was that we kick the other context and immediately delay a very short amount of time. If after that we realize the same amount is waiting go ahead and send with very low latency from leaving the client. If its ot the same and going up we can wait a bit more and batch things up at that point.

Reply:
I'm getting 12,828,553 msgs/sec with a 1ms loop instead of 10. What's your threshold for low-volume latency? I honestly think we could take this down pretty low before we start seeing it fight over CPU since `Socket#flush` with no data waiting stays completely in userland — no I/O sys calls are performed.

Thanks for the tip on `nats-server`. I was trying to consume a full core in the client. Splitting across multiple connections and publishing on a random one gets me just about there, thanks!

<img width="261" alt="Screenshot of CPU consumption of NATS server and client" src="https://user-images.githubusercontent.com/108205/77600741-22460d00-6edf-11ea-8b6e-e9fa418681dd.png">

Reply:
Again that is why I suggest we put in the backoff and check for pending to see if it was still increasing.

Reply:
[Added in the backoff](https://github.com/nats-io/nats.cr/pull/8/files#diff-200c75062e7d28e93f6906be2f4dcd13R485-R496) using the [same durations from the Rust driver](https://github.com/nats-io/nats.rs/blob/master/src/lib.rs#L324) and it was surprisingly easy. I couldn't check whether there was any data in the buffer because Crystal doesn't expose that but I was able to track it on the connection in a threadsafe way with `Atomic`.